processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -64,50 +64,48 @@
     }
 
     boolean allowReverseLookup = true;
-    int count = 1;
-    T prevVal = objects.next();
-    while (objects.hasNext()) {
-      T next = objects.next();
-      if (!(strategy.compare(prevVal, next) < 0)) {
-        allowReverseLookup = false;
-      }
-      if (prevVal instanceof Closeable) {
-        CloseQuietly.close((Closeable) prevVal);
-      }
-
-      prevVal = next;
-      ++count;
-    }
-    if (prevVal instanceof Closeable) {
-      CloseQuietly.close((Closeable) prevVal);
-    }
+    int count = 0;
 
-    ByteArrayOutputStream headerBytes = new ByteArrayOutputStream(4 + (count * 4));
+    ByteArrayOutputStream headerBytes = new ByteArrayOutputStream();
     ByteArrayOutputStream valueBytes = new ByteArrayOutputStream();
-    int offset = 0;
-
     try {
-      headerBytes.write(Ints.toByteArray(count));
+      headerBytes.write(Ints.toByteArray(0));
+
+      int offset = 0;
+      T prevVal = null;
+      do {
+        count++;
+        T next = objects.next();
+        if (allowReverseLookup && prevVal != null && !(strategy.compare(prevVal, next) < 0)) {
+          allowReverseLookup = false;
+        }
 
-      for (T object : objectsIterable) {
-        final byte[] bytes = strategy.toBytes(object);
+        final byte[] bytes = strategy.toBytes(next);
         offset += 4 + bytes.length;
         headerBytes.write(Ints.toByteArray(offset));
         valueBytes.write(Ints.toByteArray(bytes.length));
         valueBytes.write(bytes);
 
-        if (object instanceof Closeable) {
-          CloseQuietly.close((Closeable) object);
+        if (prevVal instanceof Closeable) {
+          CloseQuietly.close((Closeable) prevVal);
         }
+        prevVal = next;
+      } while (objects.hasNext());
+
+      if (prevVal instanceof Closeable) {
+        CloseQuietly.close((Closeable) prevVal);
       }
     }
     catch (IOException e) {
       throw new RuntimeException(e);
     }
+    byte[] header = headerBytes.toByteArray();
+    byte[] values = valueBytes.toByteArray();
+    System.arraycopy(Ints.toByteArray(count), 0, header, 0, 4);
 
-    ByteBuffer theBuffer = ByteBuffer.allocate(headerBytes.size() + valueBytes.size());
-    theBuffer.put(headerBytes.toByteArray());
-    theBuffer.put(valueBytes.toByteArray());
+    ByteBuffer theBuffer = ByteBuffer.allocate(header.length + values.length);
+    theBuffer.put(header);
```

Comment:
Instead of overwriting the first 4 bytes in headerBytes after the fact, you might as well just add a `theBuffer.putInt(count)` before this line and not bother having the count as part of `headerBytes`